### ISSUE TITLE
Bug in `T_HeatCapacity_Whittington` ?

### ISSUE THREAD

**Issue by albert-de-montserrat** [JuliaGeodynamics/GeoParams.jl#170]
Given 

```julia
 CharUnits_GEO = GEO_units(; viscosity=1e19, length=10km)
 x_D1 =Latent_HeatCapacity(Cp=T_HeatCapacity_Whittington())
 x_ND1 = nondimensionalize(x_D1, CharUnits_GEO)
 
 dϕdT = 0.1
 args = (dϕdT=dϕdT, T=10)
 ```

It then yields
```julia-repl
In [9]:  Cp = compute_heatcapacity(x_D1, args)
-224505.23852466405
```
@boriskaus  I guess there's a bug somewhere

**Comment by albert-de-montserrat** [comment]
My bad, I think it is correct... T has to be in K :weary:

**Comment by boriskaus** [comment]
indeed K it should be. We should probably clarify that in the help.

It should actually also work if we nondimensionalize everything and redimensionalize it after computation. Would perhaps be good to add a test to verify this, given that you now spend some time looking into this

**Comment by aelligp** [comment]
I'll address this in #169

**Comment by albert-de-montserrat** [comment]
Am I doing something wrong?

```julia
CharUnits_GEO = GEO_units(; viscosity=1e19, length=10km)
x_D =Latent_HeatCapacity(Q_L=500kJ/kg)
x_ND = nondimensionalize(x_D, CharUnits_GEO)

dϕdT = 0.1
args = (;dϕdT)
Cp = compute_heatcapacity(x_D, args)
Cp_nd = compute_heatcapacity(x_ND, args)
Cp_nd0 = nondimensionalize(Cp*(J / kg / K), CharUnits_GEO)
```

gives

```julia-repl
In [24]: dimensionalize(Cp_nd, J / kg / K, CharUnits_GEO)
1089.2726701488434 J kg^-1.0 K^-1.0
In [25]: dimensionalize(Cp_nd0, J / kg / K, CharUnits_GEO)
1100.0 J kg^-1.0 K^-1.0
```

**Comment by aelligp** [comment]
this looks like a roundoff error as: 
```
julia> CharUnits_GEO.heatcapacity
7.854534029768683e-20 m²·⁰ K⁻¹·⁰ s⁻²·⁰
```

**Comment by albert-de-montserrat** [comment]
Shouldn't 
```julia
ρCp_dim = (ρCp * CharUnits_GEO.density * CharUnits_GEO.heatcapacity)
```
be
```julia
ρCp_dim = (ρCp_ND * CharUnits_GEO.density * CharUnits_GEO.heatcapacity)
```
?

**Comment by albert-de-montserrat** [comment]
Are the [units](https://github.com/JuliaGeodynamics/GeoParams.jl/blob/05fcc3fe8183196bd4f91d18c2142d68f5764bb0/src/Energy/HeatCapacity.jl#L135) of `Q_L` correct, it's in `kJ` while `Cp` is in `J`

**Comment by boriskaus** [comment]
Isn't `kJ` automatically translated to `J`?

**Comment by albert-de-montserrat** [comment]
Nope
```julia-repl
In [16]: @unpack_val Q_L = Latent_HeatCapacity(;Q_L= 400kJ/kg);
In [17]: Q_L
400.0
In [18]: @unpack_val Q_L = Latent_HeatCapacity(;Q_L= 400e3J/kg);
In [19]: Q_L
400000.0

In [20]: compute_heatcapacity(Latent_HeatCapacity(;Q_L= 400kJ/kg), args)
1090.0
In [21]: compute_heatcapacity(Latent_HeatCapacity(;Q_L= 400e3J/kg), args)
41050.0
```

**Comment by boriskaus** [comment]
that's bad. Should be `400e3J/kg`

**Comment by albert-de-montserrat** [comment]
This issue propagates to other types:
```
# Define creep law with default values:
    # A=1.5MPa^(-n - r) / s * m^(-p)
    # E=500kJ / mol
x1 = DiffusionCreep()
T = 650 + 273.15
args = (; T=T)
TauII = 1e6
```

```julia
In [2]: ε = compute_εII(x1, TauII, args)
2.108102261290912e6
```
I guess all the default values need to be in SI

**Comment by aelligp** [comment]
> ```julia
> CharUnits_GEO = GEO_units(; viscosity=1e19, length=10km)
> x_D =Latent_HeatCapacity(Q_L=500kJ/kg)
> x_ND = nondimensionalize(x_D, CharUnits_GEO)
> 
> dϕdT = 0.1
> args = (;dϕdT)
> Cp = compute_heatcapacity(x_D, args)
> Cp_nd = compute_heatcapacity(x_ND, args)
> Cp_nd0 = nondimensionalize(Cp*(J / kg / K), CharUnits_GEO)
> ```

The problem here is that `dϕdT` is still dimensional. If we nondimensionalize this with `dϕdT_ND = nondimensionalize(dϕdT / K, CharUnits_GEO)` the values match. 
This is addressed in #169

**Comment by boriskaus** [comment]
> This issue propagates to other types:
> 
> ```
> # Define creep law with default values:
>     # A=1.5MPa^(-n - r) / s * m^(-p)
>     # E=500kJ / mol
> x1 = DiffusionCreep()
> T = 650 + 273.15
> args = (; T=T)
> TauII = 1e6
> ```
> 
>  ```julia
> In [2]: ε = compute_εII(x1, TauII, args)
> 2.108102261290912e6
> ```
> 
> I guess all the default values need to be in SI

If you add this to the `MatParams` structure, I am pretty sure that this is taken care off automatically (we convert to SI before non dimensionalisation). That is not the case if you use the values directly.

**Comment by boriskaus** [comment]
> Are the [units](https://github.com/JuliaGeodynamics/GeoParams.jl/blob/05fcc3fe8183196bd4f91d18c2142d68f5764bb0/src/Energy/HeatCapacity.jl#L135) of `Q_L` correct? it's in `kJ` while `Cp` is in `J`

Can we solve this by converting everything to SI units before non dimensionalisation?

**Comment by boriskaus** [comment]
> The problem here is that `dϕdT` is still dimensional. If we nondimensionalize this with `dϕdT_ND = nondimensionalize(dϕdT / K, CharUnits_GEO)` the values match. This is addressed in #169

that is indeed an important observation.

**Comment by albert-de-montserrat** [comment]
> > Are the [units](https://github.com/JuliaGeodynamics/GeoParams.jl/blob/05fcc3fe8183196bd4f91d18c2142d68f5764bb0/src/Energy/HeatCapacity.jl#L135) of `Q_L` correct? it's in `kJ` while `Cp` is in `J`
> 
> Can we solve this by converting everything to SI units before non dimensionalisation?

I believe `nondimensionalize` takes proper care of this. The problem is when using it directly in dimensional units, which yields the wrong result.

**Comment by aelligp** [comment]
To summarise this issue: 

There were a few things that led to errors. 
1. When dealing with the `Latent_HeatCapacity` `dϕdT` needs to be in dimensional units `[1/K]` otherwise we run into the reported error. 

> > ```julia
> > CharUnits_GEO = GEO_units(; viscosity=1e19, length=10km)
> > x_D =Latent_HeatCapacity(Q_L=500kJ/kg)
> > x_ND = nondimensionalize(x_D, CharUnits_GEO)
> > 
> > dϕdT = 0.1
> > args = (;dϕdT)
> > Cp = compute_heatcapacity(x_D, args)
> > Cp_nd = compute_heatcapacity(x_ND, args)
> > Cp_nd0 = nondimensionalize(Cp*(J / kg / K), CharUnits_GEO)
> > ```
> 
> The problem here is that `dϕdT` is still dimensional. If we nondimensionalize this with `dϕdT_ND = nondimensionalize(dϕdT / K, CharUnits_GEO)` the values match. This is addressed in #169

2. There was a problem with default values not being in SI units. This has been addressed in #173. This problem only occurs if the desired rheology struct is called outside the `MaterialParameters Struct`. If one sets up its rheology with `SetMaterialParams()`  the units are automatically converted into SI units. 
```julia
julia> Phase = SetMaterialParams(Name="Viscous Matrix",
                       Density   = ConstantDensity(),
                       CreepLaws = (PowerlawViscous(), LinearViscous(η=1e21Pa*s)))
Phase 1 : Viscous Matrix
        | [dimensional units]
        | 
        |-- Density           : Constant density: ρ=2900 kg m⁻³ 
        |-- Gravity           : Gravitational acceleration: g=9.81 m s⁻² 
        |-- CreepLaws         : Powerlaw viscosity: η0=1.0e18 Pa s, n=2.0, ε0=1.0e-15 s⁻¹  
        |                       Linear viscosity: η=1.0e21 Pa s
```
Additionally, the structure itself can be nondimensionalised as follows:
```julia
julia> CharUnits_GEO   =   GEO_units(viscosity=1e19, length=1000km);
julia> Phase = SetMaterialParams(Name="Viscous Matrix", Phase=33,
                              Density   = PT_Density(),
                              CreepLaws = (PowerlawViscous(n=3), LinearViscous(η=1e23Pa*s)),
                              CharDim   = CharUnits_GEO)
Phase 33: Viscous Matrix
        | [non-dimensional units]
        | 
        |-- Density           : P/T-dependent density: ρ0=2.9e-16, α=0.038194500000000006, β=0.01, T0=0.21454659702313156, P0=0.0 
        |-- Gravity           : Gravitational acceleration: g=9.810000000000002e18 
        |-- CreepLaws         : Powerlaw viscosity: η0=0.1, n=3, ε0=0.001  
        |                       Linear viscosity: η=10000.0 
```

**Comment by boriskaus** [comment]
Issue 2 implies that we should ensure that new material parameters that are added have default values that are in base SI units. This does _not_ apply to creep laws where we have a table with literature-derived values, because we automatically convert those values to SI units when reading in those tables.
